### ISSUE TITLE
A one-validator network freezes when non-validators are peered to it (AURA/GRANDPA)

### ISSUE THREAD

**Issue by Sushisource** [paritytech/substrate#6674]
Hello,

Using rc3 I have found that if I start a one-validator network where I have configured the chainspec such that that validator is the only authority, if I then peer non-validator nodes to it after some time (5 blocks or so), production and finalization halts until the non-validator nodes are un-peered.

With any number of validators above 1, this behavior does not happen.

Obviously this isn't problematic for real-world scenarios, but it is a bit annoying to run into if you're using one validator for simple integration testing.

This feels like a bug, as I can't find any explanation as to why this would happen anywhere in AURA/GRANDPA documentation - so if it is intentional, documenting it would be very helpful.

Thank you! Let me know if any additional details are needed.

**Comment by bkchr** [comment]
Did you checked the `dev` chain spec and how it differs to your chain spec? Because the `dev` chain spec sounds already like what you want.

**Comment by Sushisource** [comment]
@bkchr They aren't meaningfully different in any way that pops out to me. Ours is different in that we use fewer or the built in pallets than the `dev` chain does. We use AURA instead of babe. Other than that, things are very similar.

That said, I just loaded up substrate `master` (with none of my code), and it doesn't produce or finalize blocks even without adding non-validator nodes. So, that seems off.

`cargo run -- --chain dev` from the `node` directory doesn't produce blocks (purged it first):
```
 Finished dev [unoptimized + debuginfo] target(s) in 0.36s
     Running `/home/sushi/dev/work/substrate/target/debug/substrate --chain dev`
2020-07-31 16:23:18 Substrate Node
2020-07-31 16:23:18 ✌️  version 2.0.0-rc5-c32a3baa2-x86_64-linux-gnu
2020-07-31 16:23:18 ❤️  by Parity Technologies <admin@parity.io>, 2017-2020
2020-07-31 16:23:18 📋 Chain specification: Development
2020-07-31 16:23:18 🏷  Node name: useless-skirt-4441
2020-07-31 16:23:18 👤 Role: FULL
2020-07-31 16:23:18 💾 Database: RocksDb at /home/sushi/.local/share/substrate/chains/dev/db
2020-07-31 16:23:18 ⛓  Native runtime: node-257 (substrate-node-0.tx1.au10)
2020-07-31 16:23:19 💸 new validator set of size 1 has been elected via ElectionCompute::OnChain for era 0
2020-07-31 16:23:19 🔨 Initializing Genesis block/state (state: 0x17cc…dd99, header-hash: 0xa9e7…5e93)
2020-07-31 16:23:19 👴 Loading GRANDPA authority set from genesis on what appears to be first startup.
2020-07-31 16:23:20 ⏱  Loaded block-time = 3000 milliseconds from genesis on first-launch
2020-07-31 16:23:20 👶 Creating empty BABE epoch changes on what appears to be first startup.
2020-07-31 16:23:20 Using default protocol ID "sup" because none is configured in the chain specs
2020-07-31 16:23:20 🏷  Local node identity is: 12D3KooWGdqNWcvcgqGwccnmtW73X5xtiQk5yhV1sGvaDVzVuyhQ (legacy representation: QmRQWYd4JB3UL8heYwpJPLhtFxruy1ZE5P4JnkAQsEeV1S)
2020-07-31 16:23:20 block_on
2020-07-31 16:23:20 block_on
2020-07-31 16:23:20 block_on
2020-07-31 16:23:20 block_on
2020-07-31 16:23:20 block_on
2020-07-31 16:23:20 block_on
2020-07-31 16:23:20 📦 Highest known block at #0
2020-07-31 16:23:20 〽️ Prometheus server started at 127.0.0.1:9615
2020-07-31 16:23:25 💤 Idle (0 peers), best: #0 (0xa9e7…5e93), finalized #0 (0xa9e7…5e93), ⬇ 0 ⬆ 0
2020-07-31 16:23:30 💤 Idle (0 peers), best: #0 (0xa9e7…5e93), finalized #0 (0xa9e7…5e93), ⬇ 0 ⬆ 0
2020-07-31 16:23:35 💤 Idle (0 peers), best: #0 (0xa9e7…5e93), finalized #0 (0xa9e7…5e93), ⬇ 0 ⬆ 0
2020-07-31 16:23:40 💤 Idle (0 peers), best: #0 (0xa9e7…5e93), finalized #0 (0xa9e7…5e93), ⬇ 0 ⬆ 0
2020-07-31 16:23:45 💤 Idle (0 peers), best: #0 (0xa9e7…5e93), finalized #0 (0xa9e7…5e93), ⬇ 0 ⬆ 0
2020-07-31 16:23:50 💤 Idle (0 peers), best: #0 (0xa9e7…5e93), finalized #0 (0xa9e7…5e93), ⬇ 0 ⬆ 0
```

**Comment by bkchr** [comment]
@Sushisource you need to use `--dev` and not `--chain dev`.

When you just run `--chain dev`, you need to use `--chain dev --validator --alice`

**Comment by Sushisource** [comment]
@bkchr Thanks for the info. Sorry, just now getting back around to this.

So, indeed, if I run the first node with `--chain dev --validator --alice` that works perfectly fine by itself.

I then started `--chain dev --bob`, and that peered and everything continued along just fine.

In order to confirm this wasn't a bug in rc3 that has since been fixed, I synced back to the `v2.0.0-rc3` tag and re-ran the same setup. Indeed, it appears to be a bug that has since been fixed. The output from alice:

```
...
 🙌 Starting consensus session on top of parent 0x7a7e185dcb39d921da8d1adc19deef1979bc5817a83923ea073a6a1db7084720
2020-08-11 11:40:45 🎁 Prepared block for proposing at 17 [hash: 0x0eb34094147fe4efe6f298c75fe15a3e5549bbe69f4e419d65c817dc50a25d0b; parent_hash: 0x7a7e…4720; extrinsics (2): [0x941d…4567, 0x42a7…1e2a]]
2020-08-11 11:40:45 🔖 Pre-sealed block for proposal at 17. Hash now 0x6f27abf0e0a0d4d5c851a8ab94a196ea863e76a2b77420254154205365a5fef7, previously 0x0eb34094147fe4efe6f298c75fe15a3e5549bbe69f4e419d65c817dc50a25d0b.
2020-08-11 11:40:45 ✨ Imported #17 (0x6f27…fef7)
2020-08-11 11:40:46 💤 Idle (0 peers), best: #17 (0x6f27…fef7), finalized #14 (0x2847…e9d9), ⬇ 2.0kiB/s ⬆ 2.4kiB/s
2020-08-11 11:40:48 🙌 Starting consensus session on top of parent 0x6f27abf0e0a0d4d5c851a8ab94a196ea863e76a2b77420254154205365a5fef7
2020-08-11 11:40:48 🎁 Prepared block for proposing at 18 [hash: 0x04a9b1e2df6aa94ffb1b84fb9d82d1ad122fb152874eabd39fa7d30ee3aa5a7e; parent_hash: 0x6f27…fef7; extrinsics (2): [0x14de…08f7, 0xbe65…80d6]]
2020-08-11 11:40:48 🔖 Pre-sealed block for proposal at 18. Hash now 0x48a6b038f6dc45744f2cf3fb2764d2cf89766434fad134e07349daf6ec39da29, previously 0x04a9b1e2df6aa94ffb1b84fb9d82d1ad122fb152874eabd39fa7d30ee3aa5a7e.
2020-08-11 11:40:48 ✨ Imported #18 (0x48a6…da29)
2020-08-11 11:40:51 ⚙️  Syncing  0.1 bps, target=#17 (1 peers), best: #18 (0x48a6…da29), finalized #16 (0x7a7e…4720), ⬇ 0.8kiB/s ⬆ 1.4kiB/s
2020-08-11 11:40:56 ⚙️  Syncing  0.0 bps, target=#17 (1 peers), best: #18 (0x48a6…da29), finalized #16 (0x7a7e…4720), ⬇ 0.3kiB/s ⬆ 0.6kiB/s
2020-08-11 11:41:01 ⚙️  Syncing  0.0 bps, target=#17 (1 peers), best: #18 (0x48a6…da29), finalized #16 (0x7a7e…4720), ⬇ 0.3kiB/s ⬆ 0.6kiB/s
2020-08-11 11:41:06 ⚙️  Syncing  0.0 bps, target=#17 (1 peers), best: #18 (0x48a6…da29), finalized #16 (0x7a7e…4720), ⬇ 0.4kiB/s ⬆ 0.7kiB/s
2020-08-11 11:41:11 ⚙️  Syncing  0.0 bps, target=#17 (1 peers), best: #18 (0x48a6…da29), finalized #16 (0x7a7e…4720), ⬇ 0.3kiB/s ⬆ 0.6kiB/s
...
```

From bob:

```
...
📦 Highest known block at #0
2020-08-11 11:40:42 Using default protocol ID "sup" because none is configured in the chain specs
2020-08-11 11:40:42 🏷  Local node identity is: 12D3KooWMh1BfpxZknX4m4vtci4c7MsdGvgbHJiyZndiMDkmsRQw (legacy representation: QmWu48Ks4DAvM3cYd6fHRg4g3PRZ2Tw263PtV6YTz8XenN)
2020-08-11 11:40:42 Unable to bind RPC server to 127.0.0.1:9944. Trying random port.
2020-08-11 11:40:42 👶 Starting BABE Authorship worker
2020-08-11 11:40:42 📪 Libp2p listener () closed: Address already in use (os error 98)
2020-08-11 11:40:42 📪 Libp2p listener () closed: Address already in use (os error 98)
2020-08-11 11:40:45 ✨ Imported #17 (0x6f27…fef7)
2020-08-11 11:40:47 💤 Idle (0 peers), best: #17 (0x6f27…fef7), finalized #14 (0x2847…e9d9), ⬇ 2.4kiB/s ⬆ 2.0kiB/s
2020-08-11 11:40:52 💤 Idle (1 peers), best: #18 (0x48a6…da29), finalized #16 (0x7a7e…4720), ⬇ 1.4kiB/s ⬆ 0.9kiB/s
2020-08-11 11:40:57 💤 Idle (1 peers), best: #18 (0x48a6…da29), finalized #16 (0x7a7e…4720), ⬇ 0.6kiB/s ⬆ 0.3kiB/s
2020-08-11 11:41:02 💤 Idle (1 peers), best: #18 (0x48a6…da29), finalized #16 (0x7a7e…4720), ⬇ 0.6kiB/s ⬆ 0.3kiB/s
2020-08-11 11:41:07 💤 Idle (1 peers), best: #18 (0x48a6…da29), finalized #16 (0x7a7e…4720), ⬇ 0.7kiB/s ⬆ 0.4kiB/s
2020-08-11 11:41:12 💤 Idle (1 peers), best: #18 (0x48a6…da29), finalized #16 (0x7a7e…4720), ⬇ 0.6kiB/s ⬆ 0.3kiB/s
...
```

So, looks like there's nothing remaining to be done here. Thanks!
